DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c

```diff
@@ -0,0 +1,344 @@
+
+  Called by the Dynamic Table Manager, this function invokes the
+  Configuration Manager protocol interface to get the required hardware
+  information for generating the ACPI table.
+
+  If this function allocates any resources then they must be freed
+  in the FreeTpm2TableResources function.
+
+  @param [in]  This           Pointer to the table generator.
+  @param [in]  AcpiTableInfo  Pointer to the ACPI Table Info.
+  @param [in]  CfgMgrProtocol Pointer to the Configuration Manager
+                              Protocol Interface.
+  @param [out] Table          Pointer to the constructed ACPI Table.
+
+  @retval EFI_SUCCESS           Table generated successfully.
+  @retval EFI_INVALID_PARAMETER A parameter is invalid.
+  @retval EFI_NOT_FOUND         The required object was not found.
+  @retval EFI_BAD_BUFFER_SIZE   The size returned by the Configuration
+                                Manager is less than the Object size for the
+                                requested object.
+  @retval EFI_OUT_OF_RESOURCES  Memory allocation failed.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+BuildTpm2Table (
+  IN  CONST ACPI_TABLE_GENERATOR                  *CONST  This,
+  IN  CONST CM_STD_OBJ_ACPI_TABLE_INFO            *CONST  AcpiTableInfo,
+  IN  CONST EDKII_CONFIGURATION_MANAGER_PROTOCOL  *CONST  CfgMgrProtocol,
+  OUT       EFI_ACPI_DESCRIPTION_HEADER          **CONST  Table
+  )
+{
+  EFI_STATUS                  Status;
+  UINT32                      TableSize;
+  CM_ARM_TPM2_INTERFACE_INFO  *TpmInfo;
+  EFI_TPM2_ACPI_TABLE_V4      *Tpm2;
+
+  *Table = NULL;
+
+  ASSERT (
+    (This != NULL) &&
+    (AcpiTableInfo != NULL) &&
+    (CfgMgrProtocol != NULL) &&
+    (Table != NULL) &&
+    (AcpiTableInfo->TableGeneratorId == This->GeneratorID) &&
+    (AcpiTableInfo->AcpiTableSignature == This->AcpiTableSignature)
+    );
+
+  if ((AcpiTableInfo->AcpiTableRevision < This->MinAcpiTableRevision) ||
+      (AcpiTableInfo->AcpiTableRevision > This->AcpiTableRevision))
+  {
+    DEBUG ((
+      DEBUG_ERROR,
+      "ERROR: TPM2: Requested table revision = %d is not supported. "
+      "Supported table revisions: Minimum = %d. Maximum = %d\n",
+      AcpiTableInfo->AcpiTableRevision,
+      This->MinAcpiTableRevision,
+      This->AcpiTableRevision
+      ));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = GetEArmObjTpm2InterfaceInfo (
+             CfgMgrProtocol,
+             CM_NULL_TOKEN,
+             &TpmInfo,
+             NULL
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Failed to get TPM interface CM Object %r\n",
+      __func__,
+      Status
+      ));
+    return Status;
+  }
+
+  // Calculate the size of the TPM2 table
+  switch (AcpiTableInfo->AcpiTableRevision) {
+    case EFI_TPM2_ACPI_TABLE_REVISION_3:
+      TableSize = sizeof (EFI_TPM2_ACPI_TABLE);
+      break;
+
+    case EFI_TPM2_ACPI_TABLE_REVISION_4:
+      TableSize = sizeof (EFI_TPM2_ACPI_TABLE_V4);

```

Comment:
Cf: TCG Hardware Interface Description Table Format for TPM 2.0

```
The length of this table starting from the Signature field up to
and including the platform specific parameters field. (52 +
size of platform specific parameters)
```

Apparently it is possible to infer the number of bytes used based in the "Start Method Specific Parameters" based on the value of "Start Method", but I didn't find how to do that exactly:
```
The content of the Start Method specific parameters is
determined by the Start Method used by the system’s TPM
device interface. This field contains values that may be used
to initiate command processing.
If the Start Method value is 2, then this field is at least four
bytes in size and the first four bytes must be all zero.
If the Start Method value is 11 then this field is 12 bytes in
size and is described in Table 9.
If LAML and LASA are present, then this field is 12 bytes in
size.
```

As the number of bytes used seems to be variable for one value of the Start Method, I think you should add another field in CM_ARM_TPM2_INTERFACE_INFO to specify how many bytes of StartMethodParameters[12] will be used.

Reply:
Good point. At the first thought, I guess we would need to add a size field for StartMethodParameters[]. But let me think about it a little more.